src/wp-includes/abilities-api/class-wp-ability.php

```diff
@@ -488,9 +488,26 @@ public function validate_input( $input = null ) {
 					$valid_input->get_error_message()
 				)
 			);
+		} else {
+			$is_valid = true;
 		}
 
-		return true;
+		/**
+		 * Filters the input validation result for an ability.
+		 *
+		 * Allows developers to add custom validation logic on top of the default
+		 * JSON Schema validation. If default validation already failed, the filter
+		 * receives the WP_Error object and can add additional error information or
+		 * override it. If default validation passed, the filter can add additional
+		 * validation checks and return a WP_Error if those checks fail.
+		 *
+		 * @since 7.0.0
+		 *
+		 * @param true|WP_Error $is_valid      The validation result from default validation.
+		 * @param mixed         $input         The input data being validated.
+		 * @param string        $ability_name  The name of the ability.
+		 */
+		return apply_filters( 'wp_ability_validate_input', $is_valid, $input, $this->name );
```

Comment:
I'm wary of plugins doing the right thing here. This would be safer:

Consider the similar logic in the Customizer:

https://github.com/WordPress/wordpress-develop/blob/d676a07f9242d98f8d28c4d0064543a58560e304/src/wp-includes/class-wp-customize-setting.php#L602-L622

So think this could rather be:


```suggestion
		$validity = apply_filters( 'wp_ability_validate_input', $is_valid, $input, $this->name );
		if ( is_wp_error( $validity ) && ! $validity->has_errors() ) {
			$validity = true;
		}
		return $validity;
```

Reply:
Good catch.

> Or maybe returning false should result in a generic invalidity WP_Error being created.

That seems more straightforward to me; do you see any downsides to that?

Reply:
I don't see any downsides. Ideally a plugin would provide an error message, but better to err on the side of marking a non-`true` response as being an error than to accidentally let something which a plugin intends to be an error to be considered valid.

Reply:
I've applied your suggestion and explicitly checked for `false === $validity` to prevent a situation where a plugin returns `false` and it is converted to `true`.